File: app/src/test/java/djordjeh/architecture/mvp/data/TaskRepositoryTest.java
package djordjeh.architecture.mvp.data;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.data.source.TaskDataSource;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class TaskRepositoryTest {

    // https://github.com/googlesamples/android-architecture/tree/todo-mvp-rxjava/todoapp

    private static final Task task1 = new Task(1, "Title1", "Description1", false);
    private static final Task task2 = new Task(2, "Title2", "Description2", true);
    private static final List<Task> TASKS = new ArrayList<Task>() {{
        add(task1);
        add(task2);
    }};

    private TestObserver<List<Task>> testObserver;

    @Mock
    private TaskDataSource localDataSource;

    @Mock
    private TaskDataSource remoteDataSource;

    private TaskRepository repository;

    @Before
    public void setupRepository() {
        MockitoAnnotations.initMocks(this);

        repository = new TaskRepository(localDataSource, remoteDataSource);

        testObserver = new TestObserver<>();
    }

    @Test
    public void loadTasks__repositoryAfterFirstSubscription_fromRemoteSourceAndSaveThem() {
        final boolean forceUpdate = true;
        // We are fetching data from remote data source which has result
        setTasksAvailable(remoteDataSource, forceUpdate, TASKS);

        TestObserver<List<Task>> testObserver = new TestObserver<>();
        repository.tasks(forceUpdate).subscribe(testObserver);

        // Then tasks were only requested once from remote and saved by local sources
        verify(remoteDataSource).tasks(forceUpdate);
        verify(localDataSource, never()).tasks(forceUpdate);
        verify(localDataSource).save(task1);
        verify(localDataSource).save(task2);

        testObserver.assertValue(TASKS);
    }

    @Test
    public void loadTasks__repositoryAfterFirstSubscription_fromLocalSource() {
        final boolean forceUpdate = false;
        // We are fetching data from local data source which has result
        setTasksAvailable(localDataSource, forceUpdate, TASKS);

        TestObserver<List<Task>> testObserver = new TestObserver<>();
        repository.tasks(forceUpdate).subscribe(testObserver);

        verify(localDataSource).tasks(forceUpdate);
        verify(remoteDataSource, never()).tasks(forceUpdate);

        testObserver.assertValue(TASKS);
    }

    @Test
    public void loadTasks_repositoryFromRemoteDataSource() {
        final boolean forceUpdate = true;
        // We are fetching data from local data source which has result
        setTasksAvailable(remoteDataSource, forceUpdate, TASKS);

        // When tasks are requested from the tasks repository
        repository.tasks(forceUpdate).subscribe(testObserver);

        // Then tasks are loaded from the local data source
        verify(remoteDataSource).tasks(forceUpdate);
        verify(localDataSource, never()).tasks(forceUpdate);
        testObserver.assertValue(TASKS);
    }

    @Test
    public void loadTasks_repositoryFromLocalDataSource() {
        final boolean forceUpdate = false;
        // We are fetching data from local data source which has result
        setTasksAvailable(localDataSource, forceUpdate, TASKS);

        // When tasks are requested from the tasks repository
        repository.tasks(forceUpdate).subscribe(testObserver);

        // Then tasks are loaded from the local data source
        verify(localDataSource).tasks(forceUpdate);
        verify(remoteDataSource, never()).tasks(forceUpdate);
        testObserver.assertValue(TASKS);
    }

    @Test
    public void loadTask_repositoryAfterFirstSubscription_fromLocalSource() {
        setTaskAvailable(localDataSource, task1);
        setTaskAvailable(remoteDataSource, task2);

        TestObserver<Task> testObserver = new TestObserver<>();
        repository.task(1).subscribe(testObserver);

        // Then tasks were only requested once from remote and saved by local sources
        verify(localDataSource).task(1);
        verify(remoteDataSource).task(1);

        testObserver.assertValue(task1);
    }

    @Test
    public void loadTask_repositoryAfterFirstSubscription_fromRemoteSource() {
        setTaskNotAvailable(localDataSource);
        setTaskAvailable(remoteDataSource, task2);

        TestObserver<Task> testObserver = new TestObserver<>();
        repository.task(1).subscribe(testObserver);

        // Then tasks were only requested once from remote and saved by local sources
        verify(localDataSource).task(1);
        verify(remoteDataSource).task(1);

        testObserver.assertValue(task2);
    }

    @Test
    public void saveTask() {
        final Task newTask = new Task(123, "New task", "new description", false);
        when(remoteDataSource.save(newTask)).thenReturn(Single.just(newTask));
        when(localDataSource.save(newTask)).thenReturn(Single.just(newTask));

        TestObserver<Task> testObserver = new TestObserver<>();
        repository.save(newTask).subscribe(testObserver);

        verify(remoteDataSource).save(newTask);
        verify(localDataSource).save(newTask);

        testObserver.assertValue(newTask);
    }

    @Test
    public void deleteTask() {
        final Task dTask = new Task(321, "Delete task", "Task for deleting", true);
        when(remoteDataSource.delete(dTask)).thenReturn(Single.just(dTask));
        when(localDataSource.delete(dTask)).thenReturn(Single.just(dTask));

        TestObserver<Task> testObserver = new TestObserver<>();
        repository.delete(dTask).subscribe(testObserver);

        verify(remoteDataSource).delete(dTask);
        verify(localDataSource).delete(dTask);

        testObserver.assertValue(dTask);
    }

    private void setTaskAvailable(TaskDataSource dataSource, Task task) {
        // don't allow the data sources to complete.
        when(dataSource.task(1)).thenReturn(Maybe.just(task));
    }

    private void setTaskNotAvailable(TaskDataSource dataSource) {
        // don't allow the data sources to complete.
        when(dataSource.task(1)).thenReturn(Maybe.empty());
    }

    private void setTasksAvailable(TaskDataSource dataSource, boolean forceUpdate, List<Task> tasks) {
        // don't allow the data sources to complete.
        when(dataSource.tasks(forceUpdate)).thenReturn(Observable.just(tasks).concatWith(Observable.never()));
    }

    private void setTasksNotAvailable(TaskDataSource dataSource, boolean forceUpdate) {
        when(dataSource.tasks(forceUpdate)).thenReturn(Observable.just(Collections.emptyList()));
    }
}

File: app/src/test/java/djordjeh/architecture/mvp/data/TaskRepositoryTest.kt
package djordjeh.architecture.mvp.data

import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

import java.util.ArrayList

import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.data.source.TaskDataSource
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.observers.TestObserver

import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

class TaskRepositoryTest {

    private lateinit var testObserver: TestObserver<List<Task>>

    @Mock
    private lateinit var localDataSource: TaskDataSource

    @Mock
    private lateinit var remoteDataSource: TaskDataSource

    private lateinit var repository: TaskRepository

    @Before
    fun setupRepository() {
        MockitoAnnotations.initMocks(this)

        repository = TaskRepository(localDataSource, remoteDataSource)

        testObserver = TestObserver()
    }

    @Test
    fun loadTasks__repositoryAfterFirstSubscription_fromRemoteSourceAndSaveThem() {
        val forceUpdate = true
        // We are fetching data from remote data source which has result
        setTasksAvailable(remoteDataSource, forceUpdate, TASKS)

        val testObserver = TestObserver<List<Task>>()
        repository.tasks(forceUpdate).subscribe(testObserver)

        // Then tasks were only requested once from remote and saved by local sources
        verify(remoteDataSource).tasks(forceUpdate)
        verify<TaskDataSource>(localDataSource, never()).tasks(forceUpdate)
        verify<TaskDataSource>(localDataSource).save(task1)
        verify<TaskDataSource>(localDataSource).save(task2)

        testObserver.assertValue(TASKS)
    }

    @Test
    fun loadTasks__repositoryAfterFirstSubscription_fromLocalSource() {
        val forceUpdate = false
        // We are fetching data from local data source which has result
        setTasksAvailable(localDataSource, forceUpdate, TASKS)

        val testObserver = TestObserver<List<Task>>()
        repository.tasks(forceUpdate).subscribe(testObserver)

        verify(localDataSource).tasks(forceUpdate)
        verify<TaskDataSource>(remoteDataSource, never()).tasks(forceUpdate)

        testObserver.assertValue(TASKS)
    }

    @Test
    fun loadTasks_repositoryFromRemoteDataSource() {
        val forceUpdate = true
        // We are fetching data from local data source which has result
        setTasksAvailable(remoteDataSource, forceUpdate, TASKS)

        // When tasks are requested from the tasks repository
        repository.tasks(forceUpdate).subscribe(testObserver)

        // Then tasks are loaded from the local data source
        verify(remoteDataSource).tasks(forceUpdate)
        verify<TaskDataSource>(localDataSource, never()).tasks(forceUpdate)
        testObserver.assertValue(TASKS)
    }

    @Test
    fun loadTasks_repositoryFromLocalDataSource() {
        val forceUpdate = false
        // We are fetching data from local data source which has result
        setTasksAvailable(localDataSource, forceUpdate, TASKS)

        // When tasks are requested from the tasks repository
        repository.tasks(forceUpdate).subscribe(testObserver)

        // Then tasks are loaded from the local data source
        verify(localDataSource).tasks(forceUpdate)
        verify<TaskDataSource>(remoteDataSource, never()).tasks(forceUpdate)
        testObserver.assertValue(TASKS)
    }

    @Test
    fun loadTask_repositoryAfterFirstSubscription_fromLocalSource() {
        setTaskAvailable(localDataSource, task1)
        setTaskAvailable(remoteDataSource, task2)

        val testObserver = TestObserver<Task>()
        repository.task(1).subscribe(testObserver)

        // Then tasks were only requested once from remote and saved by local sources
        verify(localDataSource).task(1)
        verify(remoteDataSource).task(1)

        testObserver.assertValue(task1)
    }

    @Test
    fun loadTask_repositoryAfterFirstSubscription_fromRemoteSource() {
        setTaskNotAvailable(localDataSource)
        setTaskAvailable(remoteDataSource, task2)

        val testObserver = TestObserver<Task>()
        repository.task(1).subscribe(testObserver)

        // Then tasks were only requested once from remote and saved by local sources
        verify(localDataSource).task(1)
        verify(remoteDataSource).task(1)

        testObserver.assertValue(task2)
    }

    @Test
    fun saveTask() {
        val newTask = Task(123, "New task", "new description", false)
        `when`(remoteDataSource.save(newTask)).thenReturn(Single.just(newTask))
        `when`(localDataSource.save(newTask)).thenReturn(Single.just(newTask))

        val testObserver = TestObserver<Task>()
        repository.save(newTask).subscribe(testObserver)

        verify(remoteDataSource).save(newTask)
        verify(localDataSource).save(newTask)

        testObserver.assertValue(newTask)
    }

    @Test
    fun deleteTask() {
        val dTask = Task(321, "Delete task", "Task for deleting", true)
        `when`(remoteDataSource.delete(dTask)).thenReturn(Single.just(dTask))
        `when`(localDataSource.delete(dTask)).thenReturn(Single.just(dTask))

        val testObserver = TestObserver<Task>()
        repository.delete(dTask).subscribe(testObserver)

        verify(remoteDataSource).delete(dTask)
        verify(localDataSource).delete(dTask)

        testObserver.assertValue(dTask)
    }

    private fun setTaskAvailable(dataSource: TaskDataSource, task: Task) {
        // don't allow the data sources to complete.
        `when`(dataSource.task(1)).thenReturn(Maybe.just(task))
    }

    private fun setTaskNotAvailable(dataSource: TaskDataSource) {
        // don't allow the data sources to complete.
        `when`(dataSource.task(1)).thenReturn(Maybe.empty())
    }

    private fun setTasksAvailable(dataSource: TaskDataSource, forceUpdate: Boolean, tasks: List<Task>) {
        // don't allow the data sources to complete.
        `when`(dataSource.tasks(forceUpdate)).thenReturn(Observable.just(tasks).concatWith(Observable.never()))
    }

    private fun setTasksNotAvailable(dataSource: TaskDataSource, forceUpdate: Boolean) {
        `when`(dataSource.tasks(forceUpdate)).thenReturn(Observable.just(emptyList()))
    }

    companion object {

        private val task1 = Task(1, "Title1", "Description1", false)
        private val task2 = Task(2, "Title2", "Description2", true)
        private val TASKS = object : ArrayList<Task>() {
            init {
                add(task1)
                add(task2)
            }
        }
    }
}
